filament/backend/src/vulkan/memory/ResourceCounter.h

```diff
@@ -0,0 +1,337 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TNT_FILAMENT_BACKEND_VULKAN_MEMORY_RESOURCECOUNTER_H
+#define TNT_FILAMENT_BACKEND_VULKAN_MEMORY_RESOURCECOUNTER_H
+
+#include "vulkan/memory/Resource.h"
+
+#include <utils/Mutex.h>
+#include <utils/Panic.h>
+
+#include <functional>
+
+namespace filament::backend::fvkmemory {
+
+namespace {
+constexpr bool IS_THREAD_SAFE = true;
+constexpr bool IS_NOT_THREAD_SAFE = false;
+} // anonymous namespace
+
+using DestructionFunc = std::function<void(ResourceType, HandleId)>;
+
+template<bool THREAD_SAFE>
+struct alignas(8) CounterImpl {
+public:
+    inline void inc() noexcept {
+        if constexpr (THREAD_SAFE) {
+            std::unique_lock<utils::Mutex> lock(*mMutex);
+            if UTILS_LIKELY (!mDestroyed) {
+                mCount++;
+            }
+        } else {
+            mCount++;
+        }
+    }
+
+    inline void dec() noexcept {
+        if constexpr (THREAD_SAFE) {
+            if (mDestroyed) {
+                return;
+            }
+            bool shouldDestroy = false;
+            {
+                std::unique_lock<utils::Mutex> lock(*mMutex);
+                shouldDestroy = --mCount == 0;
+            }
+            if (shouldDestroy) {
+                destroyLater(type, id);
+                mDestroyed = true;
+            }
+        } else {
+            assert_invariant(mCount > 0);
+            if (--mCount == 0) {
+                destroyNow(type, id);
+                mDestroyed = true;
+            }
+        }
+    }
+
+    CounterImpl(ResourceType type, HandleId id)
+        : id(id),
+          type(type),
+          mDestroyed(false),
+          mCount(0) {
+        if constexpr (THREAD_SAFE) {
+            mMutex = std::make_unique<utils::Mutex>();
+        }
+    }
+
+    CounterImpl()
+        : id(HandleBase::nullid),
+          type(ResourceType::UNDEFINED_TYPE),
+          mDestroyed(false),
+          mCount(0) {}
+
+    CounterImpl(CounterImpl const& rhs)
+        : id(rhs.id),
+          type(rhs.type),
+          mDestroyed(false),
+          mCount(rhs.mCount) {
+        if (rhs.mMutex) {
+            mMutex = std::make_unique<utils::Mutex>();
+        }
+    }
+
+    ~CounterImpl() = default;
+
+    HandleId const id;        // 8
+    ResourceType const type;  // 1
+
+private:
+    void destroyNow(ResourceType type, HandleId id);
+    void destroyLater(ResourceType type, HandleId id);
+
+    bool mDestroyed = false;  // 1
+    uint16_t padding = 0;     // 2
+    uint32_t mCount;          // 4
+
+    // To keep down the size of this class, we only allocate utils::Mutex if necessary.
+    std::unique_ptr<utils::Mutex> mMutex; // 8
```

Comment:
Could you do something like:

```
    typename std::conditional_t<THREAD_SAFE, utils::Mutex, std::monostate> mMutex;
```

I _think_ that only adds the `mMutex` member variable if `THREAD_SAFE` is `true`.

Reply:
this is the c++ magic i was looking for:) (but couldnt' find).

Reply:
done